Sources/Apollo/GraphQLPaginatedQueryWatcher.swift

```diff
@@ -0,0 +1,157 @@
+#if !COCOAPODS
+import ApolloAPI
+#endif
+import Foundation
+
+public typealias Cursor = String
+
+/// Handles pagination in the queue by managing multiple query watchers.
+final class GraphQLPaginatedQueryWatcher<Query: GraphQLQuery, T>: Cancellable {
+
+  public struct Page {
+    let hasNextPage: Bool
+    let endCursor: Cursor?
+  }
+
+  public struct DataResponse {
+    let allResponses: [T]
+    let mostRecent: T
+    let source: GraphQLResult<Query.Data>.Source
+  }
+
+  /// Given a page, create a query of the type this watcher is responsible for
+  public typealias CreatePageQuery = (Page) -> Query?
+
+  private typealias ResultHandler = (Result<GraphQLResult<Query.Data>, Error>) -> Void
+
+  private let client: any ApolloClientProtocol
+
+  var initialWatcher: GraphQLQueryWatcher<Query>?
+  private var subsequentWatchers: [GraphQLQueryWatcher<Query>] = []
+
+  private let createPageQuery: CreatePageQuery
+  private let nextPageTransform: (DataResponse) -> T
+
+  private var modelMap: [Cursor?: T] = [:]
+  private var cursorOrder: [Cursor?] = []
+  private var resultHandler: ResultHandler?
+  private var callbackQueue: DispatchQueue
+  private var page: Page?
+
+  /// Designated initializer
+  ///
+  /// - Parameters:
+  ///   - client: The client protocol to pass in
+  ///   - inititalCachePolicy: The preferred cache policy for the initlal page. Defaults to `returnCacheDataAndFetch`.
+  ///   - callbackQueue: The queue for response callbacks.
+  ///   - query: The query to watch
+  ///   - createPageQuery: A function which creates a new `Query` given some pagination information.
+  ///   - transform: Transforms the `Query.Data` to the intended model and extracts the `Page` from the `Data`.
+  ///   - nextPageTransform: A function which combines extant data with the new data from the next page
+  ///   - onReceiveResults: The callback function which returns changes or an error.
+  public init(
+    client: ApolloClientProtocol,
+    inititalCachePolicy: CachePolicy = .returnCacheDataAndFetch,
+    callbackQueue: DispatchQueue = .main,
+    query: Query,
+    createPageQuery: @escaping CreatePageQuery,
+    transform: @escaping (Query.Data) -> (T?, Page?)?,
```

Comment:
Including a transform here _may_ seem opinionated, but it's also the easiest way to avoid the hairy matter of returning several `Query.Data` for the user to combine themselves (or worse -- trying to amalgamate them ourselves!) ]

I figured it's best to have them just merge their formed models instead of assuming

Reply:
I'm a bit confused why we need both `transform` and `nextPageTransform`. What are each of these things achieving individually? I'd like to consider if there is a way to combine these to make the API easier for users to consume.

Reply:
I suppose part of the confusion is naming -- 

`transform`: Transforms a `Query.Data` to some user-defined object.
`nextPageTransform`: Should probably be renamed to `mergeStrategy` or something -- as the idea behind it is that it gives you three things to play with in order to construct your canonical "merged" result: 
1. `allData`: The constructed model of each individual page
2. `mostRecentChange`: The constructed model that represents what's new.
3. `source`: whether that change came from the cache or the server. Useful for both debugging + I could imagine use-cases where you may want to do something differently based on the source. 


The underlying thought behind this is that operation on the level of `Query.Data` may open up a can of worms (as you mentioned in your review); I cannot make assumptions of how best to merge `some Query.Data`s together. My thoughts on it were predicated on the need to have this be type-safe, and the need to avoid having to declare a separate cache mutation for each paginated query. If we were to compromise on type safety, it requires too much knowledge of the underlying `__data` and could cause run-time errors if the code-gen changes (as we would be passing string keys around). If we were to compromise and declare cache mutations for each paginated query, it scales too poorly for large codebases and introduces a certain amount of extra generated code that I fear may un-do some of the generated output size wins of 1.x.